File: hawk/src/test/java/com/orhanobut/hawk/HawkConverterTest.java
package com.orhanobut.hawk;

import com.google.gson.Gson;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import static com.google.common.truth.Truth.assertThat;
import static junit.framework.Assert.fail;

@RunWith(RobolectricTestRunner.class)
@Config(constants = BuildConfig.class, sdk = 18)
public class HawkConverterTest {

  private Converter converter;
  private Parser parser;
  private Serializer serializer;

  static class Foo {
  }

  @Before public void setup() {
    parser = new GsonParser(new Gson());
    converter = new HawkConverter(parser);
    serializer = new HawkSerializer(new LogInterceptor() {
      @Override public void onLog(String message) {
        // ignore
      }
    });
  }

  @Test public void createInstanceWithInvalidValues() {
    try {
      new HawkConverter(null);
      fail();
    } catch (Exception e) {
      assertThat(e).hasMessage("Parser should not be null");
    }
  }

  @Test public void encodeInvalidValues() {
    assertThat(converter.toString(null)).isNull();
  }

  @Test public void encodeString() {
    String text = "text";
    String expected = parser.toJson(text);
    String actual = converter.toString(text);

    assertThat(actual).isEqualTo(expected);
  }

  @Test public void encodeCustomObject() {
    Foo data = new Foo();
    String expected = parser.toJson(data);
    String actual = converter.toString(data);

    assertThat(actual).isEqualTo(expected);
  }

  @Test public void encodeList() {
    List<String> data = new ArrayList<>();
    data.add("test");
    String expected = parser.toJson(data);
    String actual = converter.toString(data);

    assertThat(actual).isEqualTo(expected);
  }

  @Test public void encodeMap() {
    Map<String, String> data = new HashMap<>();
    data.put("key", "value");
    String expected = parser.toJson(data);
    String actual = converter.toString(data);

    assertThat(actual).isEqualTo(expected);
  }

  @Test public void encodeSet() {
    Set<String> data = new HashSet<>();
    data.add("key");
    String expected = parser.toJson(data);
    String actual = converter.toString(data);

    assertThat(actual).isEqualTo(expected);
  }

  @Test public void decodeInvalidValues() throws Exception {
    assertThat(converter.fromString(null, null)).isNull();
    try {
      assertThat(converter.fromString("value", null)).isNull();
      fail();
    } catch (Exception e) {
      assertThat(e).hasMessage("data info should not be null");
    }
  }

  @Test public void decodeObject() throws Exception {
    String clazz = "java.lang.String";
    String info = "00V";
    String cipher = "cipher";
    DataInfo dataInfo = serializer.deserialize(clazz + "##" + info + "@" + cipher);
    String actual = converter.fromString(cipher, dataInfo);
    assertThat(actual).isEqualTo(cipher);
  }

}

File: hawk/src/test/java/com/orhanobut/hawk/HawkConverterTest.kt
package com.orhanobut.hawk

import com.google.gson.Gson

import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

import java.util.ArrayList
import java.util.HashMap
import java.util.HashSet

import com.google.common.truth.Truth.assertThat
import junit.framework.Assert.fail

@RunWith(RobolectricTestRunner::class)
class HawkConverterTest {

  private lateinit var converter: Converter
  private lateinit var parser: Parser
  private lateinit var serializer: Serializer

  internal class Foo

  @Before fun setup() {
    parser = GsonParser(Gson())
    converter = HawkConverter(parser)
    serializer = HawkSerializer(LogInterceptor {
      // ignore
    })
  }

  @Test fun createInstanceWithInvalidValues() {
    try {
      HawkConverter(null)
      fail()
    } catch (e: Exception) {
      assertThat(e).hasMessage("Parser should not be null")
    }

  }

  @Test fun encodeInvalidValues() {
    assertThat(converter.toString<Any>(null)).isNull()
  }

  @Test fun encodeString() {
    val text = "text"
    val expected = parser.toJson(text)
    val actual = converter.toString(text)

    assertThat(actual).isEqualTo(expected)
  }

  @Test fun encodeCustomObject() {
    val data = Foo()
    val expected = parser.toJson(data)
    val actual = converter.toString(data)

    assertThat(actual).isEqualTo(expected)
  }

  @Test fun encodeList() {
    val data = ArrayList<String>()
    data.add("test")
    val expected = parser.toJson(data)
    val actual = converter.toString<List<String>>(data)

    assertThat(actual).isEqualTo(expected)
  }

  @Test fun encodeMap() {
    val data = HashMap<String, String>()
    data["key"] = "value"
    val expected = parser.toJson(data)
    val actual = converter.toString<Map<String, String>>(data)

    assertThat(actual).isEqualTo(expected)
  }

  @Test fun encodeSet() {
    val data = HashSet<String>()
    data.add("key")
    val expected = parser.toJson(data)
    val actual = converter.toString<Set<String>>(data)

    assertThat(actual).isEqualTo(expected)
  }

  @Test @Throws(Exception::class)
  fun decodeInvalidValues() {
    assertThat(converter.fromString<Any>(null, null)).isNull()
    try {
      assertThat(converter.fromString<Any>("value", null)).isNull()
      fail()
    } catch (e: Exception) {
      assertThat(e).hasMessage("data info should not be null")
    }

  }

  @Test @Throws(Exception::class)
  fun decodeObject() {
    val clazz = "java.lang.String"
    val info = "00V"
    val cipher = "cipher"
    val dataInfo = serializer.deserialize("$clazz##$info@$cipher")
    val actual = converter.fromString<String>(cipher, dataInfo)
    assertThat(actual).isEqualTo(cipher)
  }

}
